File: 17-单例OC/17-单例OC/NetworkTools.m
//
//  NetworkTools.m
//  17-单例OC
//
//  Created by Wei Zhou on 03/10/2016.
//  Copyright © 2016 Wei Zhou. All rights reserved.
//

#import "NetworkTools.h"

@implementation NetworkTools

+ (instancetype) shareNetworkTools {
    static NetworkTools *_instance;
    // onceToken默认等于0, 只要执行一次之后就不等于0了, 原理是通过判断onceToken是否等于0决定是否需要执行block
    static dispatch_once_t onceToken;
    
    NSLog(@"\nHalo %ld", onceToken);
    
    dispatch_once(&onceToken, ^{
        _instance = [[NetworkTools alloc] init];
    });
    
    return _instance;
}

@end

File: 16-单例/16-单例/NetworkTools.swift
//
//  NetworkTools.swift
//  16-单例
//
//  Created by Wei Zhou on 03/10/2016.
//  Copyright © 2016 Wei Zhou. All rights reserved.
//

import UIKit

class NetworkTools: NSObject {
    
//    static let _instance: NetworkTools = NetworkTools()
//    
//    class func shareNetworkTools() -> NetworkTools {
//        return _instance
//    }
//    
//    override init() {
//        print("chacha")
//    }
 
    // 如果在Swift中编写单例, 推荐使用如下写法
    // 而且苹果有统一的命名规范, 但凡是单例统一是用shareInstance
    static let shareInstance: NetworkTools =  NetworkTools()
    
}
